docs/adr/0020-application-architecture.md

```diff
@@ -0,0 +1,110 @@
+# Use a documented Application Architecture for new back-end services
+
+- Deciders: Ben Bangert, (your name here)
+- Date: 2020-08-10
+
+## Context and Problem Statement
+
+The FxA back-end service stack contains an application architecture that is ad-hoc, not documented, and missing modern features (such as Dependency Injection) which results in the following problems:
+
+- New developers struggle to get up to speed as they must learn the architecture by reading the code as we have no documentation on the application structure, why they're structured the way they are, or how new components should be added to fit in. Each back-end service may vary in its ad-hoc architecture as well.
+- Adding new objects needed in a route handler can be time-consuming as the object must be plumbed through the entire initialization chain vs. more elegant methods like Dependency Injection (DI).
+- Not clear where/how to add new components and takes time to study/understand how things are currently setup in an attempt to mimic the structure for the new component.
+- Time consuming to setup boiler-plate for components, as we have no tooling to work with the current ad-hoc application architectures.
+- Our ad-hoc architecture frequently mixes concerns such as having business logic mixed in with request handling logic, and has other warts from its evolution over time vs. being planned up front.
+- New back-end services evolve differently resulting in more ad-hoc application architectures to learn.
+- Shared components in `fxa-shared` can't rely on basic object lifecycles or setup approaches as they may be used in multiple different ad-hoc application architectures.
+
+Not choosing an application framework means that we have choosen to make ad-hoc application architectures which will continue to exhibit the problems above.
+
+It is assumed that the four newest FxA back-end services (admin-server, support-panel, event-broker, gql-api) will be switched to the chosen approach for consistency.
```

Comment:
I wonder how we could best approach transitioning our mature backend services. I don't think there's an overwhelming reason to transition them immediately, and in some cases it might make sense to not switch them at all, instead allowing them to exist until FxA architecture changes and they are retired. But waiting too long risks the new approach just adding to our list of varying tech stacks we have to learn and look after, especially if it's de-prioritized.

Reply:
I believe we can transition some of the existing services partly by moving the business logic into shared components that something like gql-api can use. This can at least reduce the amount of code in the legacy systems which should make it more manageable.

Reply:
@jodyheavener see response I wrote to vijay, as it occurred to me how to backport some of the conventions into the mature services.

Reply:
Thanks for the followup! I would be happy to be involved in one of the small switches (even if just observing or reviewing PRs).